js/views/modals/purchase/Purchase.js

```diff
@@ -123,40 +107,84 @@ export default class extends BaseModal {
   }
 
   changeQuantityInput(e) {
-    this.order.get('items').at(0).set('quantity', $(e.target).val());
+    this.order.get('items').at(0).set(this.getFormData($(e.target)));
   }
 
-  clickPayBtn() {
-    this.$confirmPay.removeClass('hide');
+  clickNewAddress() {
+    launchSettingsModal({ initTab: 'Addresses' });
   }
 
-  clickConfirmBtn() {
-    // confirm the purchase
-    this.purchaseListing();
-  }
-
-  closeConfirmPay() {
-    this.$confirmPay.addClass('hide');
+  updateShippingOption(opts) {
+    // set the shipping option
+    const oShipping = this.order.get('items').at(0).get('shipping');
+    oShipping.set({ name: opts.name, service: opts.service });
+    this.actionBtn.render();
   }
 
   purchaseListing() {
+    // clear any old errors
+    const allErrContainers = this.$('div[class $="-errors"]');
+    allErrContainers.html('');
+    let shippingError = false;
+
+    // if the listing has shipping, and a shipping option has been selected, set it
+    if (this.listing.get('shippingOptions') && this.listing.get('shippingOptions').length) {
+      if (this.shipping.selectedAddress && !!this.order.get('items').at(0).get('shipping')
+          .get('name')) {
+        // set the address
+        this.order.addAddress(this.shipping.selectedAddress);
+      } else {
+        this.insertErrors(this.$shippingErrors, [app.polyglot.t('purchase.errors.missingAddress')]);
+        shippingError = true;
+      }
+    }
+
     // set the moderator
-    this.order.set('moderator', this.moderators.selectedIDs[0]);
-
-    $.post({
-      url: app.getServerUrl('ob/purchase'),
-      data: JSON.stringify(this.order.toJSON()),
-    })
-      .done((data) => {
-        console.log(data);
-      })
-      .fail((data) => {
-        console.log(data);
+    this.order.set({ moderator: this.moderators.selectedIDs[0] }, { validate: true });
+
+    if (!this.order.validationError) {
+      if (!shippingError) {
+        $.post({
+          url: app.getServerUrl('ob/purchase'),
+          data: JSON.stringify(this.order.toJSON()),
+          dataType: 'json',
+          contentType: 'application/json',
+        })
+          .done((data) => {
+            this.state.phase = 'pending';
+            this.actionBtn.render();
+            console.log(data);
+          })
+          .fail((jqXHR) => {
+            const errMsg = jqXHR.responseJSON ? jqXHR.responseJSON.reason : '';
```

Comment:
I proposed we use this:
`errMsg = jqXhr.responseJSON && jqXhr.responseJSON.reason || '';`

and you put in this:
`errMsg = jqXHR.responseJSON ? jqXHR.responseJSON.reason : '';`

It's not a massive deal, but the benefit of the first version is that if responseJSON or reason are not defined, it will set the error message to an empty string.

In the second version, if reason is not defined, you will set the error message to undefined. I think openSimpleMessage will handle that gracefully, so it's up to you if you want to change this.

FWIW, 'reason' is a custom field provided by the server. If the request fails before getting to the server, reason won't be there.